### PR TITLE
[cifmw_cephadm] Enable the usage of custom `haproxy` and `keepalived` container images

### PR DESCRIPTION
**Pull request by rebtoor**
This PR replace https://github.com/openstack-k8s-operators/ci-framework/pull/765

As a pull request owner and reviewers, I checked that:
- [x] Appropriate testing is done and actually running
- [x] Appropriate documentation exists and/or is up-to-date:
  - [x] README in the role
  - [x] Content of the docs/source is reflecting the changes

### CODE REVIEW

File: ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml [hunk 1]
```diff
@@ -47,6 +47,25 @@
     {{ cifmw_cephadm_container_ns }}/{{ cifmw_cephadm_container_image }}
   changed_when: false
 
+- name: Set haproxy container image in ceph configuration
+  when:
+    - cifmw_cephadm_haproxy_container_image is defined
+    - not (cifmw_cephadm_default_container | default(false) | bool)
+  become: true
+  ansible.builtin.command: |
```
Comment (pablintino) [inline ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:55]: Just a style suggestion. If you change the yaml pipe `|` to `>-` you can get rid of `\` at the very end of each line.

Comment (rebtoor) [reply ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:55]: yeah, indeed... i left the block scalar style of the existing tasks for coherence. IMHO i'd like to propose a refactor of this tasks file (using a dict to define custom container images) so i will take care of style issues there if you don't mind. :)

**Comment by fmount** [comment]
+1 from me, I'll let @fultonj to look and approve the PR

**Review by fultonj — APPROVED** [review]
+2

**Comment by fultonj** [comment]
[podified-multinode-hci-deployment-crc](https://review.rdoproject.org/zuul/build/bac23b2a8a0b42a68cc04b4b814430f5)

failed as seen here:

https://logserver.rdoproject.org/14/714/b6baee16afe9fecb049362f0b5a95ab8251f9d57/github-check/podified-multinode-hci-deployment-crc/931c000/controller/ci-framework-data/logs/crc/pods/openstack_dataplane-deployment-nova-custom-ceph-openstack-edpm-28shq_6ff6d52d-047e-48a3-97cd-86c944c2a905/openstackansibleee/0.log

an explanation is here:

https://github.com/openstack-k8s-operators/ci-framework/pull/714#issuecomment-1779462543

thus, we need a patch to ci-framework to create the missing secret and this needs to be rebased on it. It's been on my todo list.

**Comment by fultonj** [comment]
I'm hoping https://github.com/openstack-k8s-operators/ci-framework/pull/777 will fix the HCI job and then this can be rebased on it.

**Comment by fultonj** [comment]
@rebtoor to make the hci job green you can update `ci_framework/roles/hci_prepare/templates/dpservice-nova-custom-ceph.yml.j2` to have the migration ssh key. Or you can rebase after https://github.com/openstack-k8s-operators/ci-framework/pull/714 merges.

```
  secrets:
    - nova-cell1-compute-config
    - nova-migration-ssh-key
```

**Comment by fultonj** [comment]
Try rebasing against main and I expect the HCI test to pass.

**Comment by rebtoor** [comment]
@fultonj done. Feel free to lgtm again :)

**Comment by fultonj** [comment]
I see tempest failed on tempest.scenario.test_network_basic_ops.TestNetworkBasicOps 

https://logserver.rdoproject.org/73/773/729c865d3c951680207944ae3d8f798787e3bf36/github-check/podified-multinode-hci-deployment-crc/7b0cb0d/controller/ci-framework-data/tests/tempest/stestr_results.html

and that's why podified-multinode-hci-deployment-crc is red

https://review.rdoproject.org/zuul/build/7b0cb0d1673949fe80acfcabf9453185

**Comment by fultonj** [comment]
/recheck

setting the default  containers in the upstream job should not affect tempest.

File: ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml [hunk 2]
```diff
@@ -47,6 +47,25 @@
...
+    {{ cifmw_cephadm_ceph_cli }} config set mgr mgr/cephadm/container_image_haproxy \
+    {{ cifmw_cephadm_haproxy_container_image }}
+  changed_when: false
+
+- name: Set keepalived container image in ceph configuration
+  when:
+    - cifmw_cephadm_keepalived_container_image is defined
+    - not (cifmw_cephadm_default_container | default(false) | bool)
+  ansible.builtin.command: |
+    {{ cifmw_cephadm_ceph_cli }} config set mgr mgr/cephadm/container_image_keepalived \
```
Comment (jamepark4) [inline ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:65]: I think this task needs a `become: true`, @fyanac tried this downstream and was seeing permission issues accessing the spec yaml '/home/ceph-admin/specs/ceph_spec.yaml' when creating the container.

Comment (fultonj) [reply ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:65]: Yes, I see `become: true` for similar tasks. It's just missing for this one.

Comment (rebtoor) [reply ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:65]: Nice catch! I splitted the tasks but i forgot to add `become: true` to the second one.

Comment (fyanac) [reply ci_framework/roles/cifmw_cephadm/tasks/cephadm_config_set.yml:66]: > I think this task needs a `become: true`, @fyanac tried this downstream and was seeing permission issues accessing the spec yaml '/home/ceph-admin/specs/ceph_spec.yaml' when creating the container.

yes, it works. Thank you :)

**Comment by fultonj** [comment]
Still failing on tempest.

https://logserver.rdoproject.org/73/773/24a10cacca06a5181b880b93af574f4597fb227e/github-check/podified-multinode-hci-deployment-crc/4f57b26/controller/ci-framework-data/tests/tempest/stestr_results.html

**Comment by fultonj** [comment]
Some tempest output on test `test_connectivity_between_vms_on_different_networks`

```
Starting network: udhcpc: started, v1.29.3
udhcpc: sending discover
udhcpc: sending select for 10.100.0.4
udhcpc: lease of 10.100.0.4 obtained, lease time 43200
route: SIOCADDRT: File exists
WARN: failed: route add -net "0.0.0.0/0" gw "10.100.0.1"
OK
...
=== network info ===
if-info: lo,up,127.0.0.1,8,,
if-info: eth0,up,10.100.0.4,28,fe80::f816:3eff:fe5a:4d7/64,
ip-route:default via 10.100.0.1 dev eth0 
ip-route:10.100.0.0/28 dev eth0 scope link  src 10.100.0.4 
ip-route:169.254.169.254 via 10.100.0.2 dev eth0 
ip-route6:fe80::/64 dev eth0  metric 256 
ip-route6:ff00::/8 dev eth0  metric 256
...
Traceback (most recent call last):
  File "/usr/lib/python3.9/site-packages/tempest/common/utils/__init__.py", line 70, in wrapper
    return f(*func_args, **func_kwargs)
  File "/usr/lib/python3.9/site-packages/tempest/scenario/test_network_basic_ops.py", line 514, in test_connectivity_between_vms_on_different_networks
    self._check_server_connectivity(self.floating_ip_tuple.floating_ip,
  File "/usr/lib/python3.9/site-packages/tempest/scenario/test_network_basic_ops.py", line 388, in _check_server_connectivity
    self.check_remote_connectivity(ssh_source, remote_ip,
  File "/usr/lib/python3.9/site-packages/tempest/scenario/manager.py", line 1465, in check_remote_connectivity
    self.fail(msg)
  File "/usr/lib64/python3.9/unittest/case.py", line 676, in fail
    raise self.failureException(msg)
AssertionError: Timed out waiting for 10.100.0.21 to become reachable from 192.168.122.209
```

**Comment by raukadah** [comment]
recheck https://softwarefactory-project.io/zuul/t/rdoproject.org/builds?job_name=podified-multinode-hci-deployment-crc&skip=0 is passing now.

**Comment by raukadah** [comment]
recheck node hold tempest debugging

**Comment by raukadah** [comment]
recheck

**Review by raukadah — APPROVED** [review]
/approve

**Pull request merged**